src/media_tools/av_parsers.c

```diff
@@ -1545,6 +1545,595 @@ GF_Err gf_m4a_write_config(GF_M4ADecSpecInfo *cfg, char **dsi, u32 *dsi_size)
+		u8 leb128_byte = gf_bs_read_u8(bs);
+		value |= ((leb128_byte & 0x7f) << (i * 7));
+		Leb128Bytes += 1;
+		if (!(leb128_byte & 0x80)) {
+			break;
+		}
+	}
+
+	if (opt_Leb128Bytes) {
+		*opt_Leb128Bytes = Leb128Bytes;
+	}
+	return value;
+}
+
+#define OBU_COPY \
+	gf_bs_seek(bs, pos); \
+	GF_AV1_OBUArrayEntry *a; \
+	GF_SAFEALLOC(a, GF_AV1_OBUArrayEntry); \
+	a->obu = gf_malloc((size_t)obu_length); \
+	gf_bs_read_data(bs, a->obu, (u32)obu_length); \
+	a->obu_length = obu_length; \
+	a->obu_type = obu_type;
+
+static void av1_populate_state_from_obu(GF_BitStream *bs, u64 pos, u64 obu_length, ObuType obu_type, AV1State *state)
+{
+	if (av1_is_obu_header(obu_type)) {
+		OBU_COPY;
+		if (!state->frame_state.header_obus) state->frame_state.header_obus = gf_list_new();
+		gf_list_add(state->frame_state.header_obus, a);
+	}
+	if (av1_is_obu_frame(obu_type)) {
+		OBU_COPY;
+		if (!state->frame_state.frame_obus) state->frame_state.frame_obus = gf_list_new();
+		gf_list_add(state->frame_state.frame_obus, a);
+	}
+}
+
+GF_Err aom_av1_parse_obu_from_section5(GF_BitStream *bs, AV1State *state) {
+	ObuType obu_type = OBU_SKIP;
+	while (obu_type != OBU_TEMPORAL_DELIMITER && gf_bs_available(bs)) {
+		u64 pos = gf_bs_get_position(bs), obu_length = 0;
+		GF_Err e;
+
+		e = gf_media_aom_av1_parse_obu(bs, &obu_type, &obu_length, state);
+		if (e) return e;
+		assert(obu_length == gf_bs_get_position(bs) - pos);
+		
+		GF_LOG(GF_LOG_DEBUG, GF_LOG_CONTAINER, ("[AV1] Section5 OBU detected (size "LLU")\n", obu_length));
+		av1_populate_state_from_obu(bs, pos, obu_length, obu_type, state);
+	}
+
+	return GF_OK;
+}
+
+GF_Err aom_av1_parse_obu_from_annexb(GF_BitStream *bs, AV1State *state)
+{
+	GF_Err e = GF_OK;
+	assert(bs && state);
+	u64 sz = leb128(bs, NULL);
+	GF_LOG(GF_LOG_DEBUG, GF_LOG_CONTAINER, ("[AV1] Annex B temporal unit detected (size "LLU") ***** \n", sz));
+	while (sz > 0) {
+		u8 Leb128Bytes = 0;
+		u64 frame_unit_size = leb128(bs, &Leb128Bytes);
+		if (sz < Leb128Bytes + frame_unit_size) {
+			GF_LOG(GF_LOG_ERROR, GF_LOG_CODING, ("[AV1] Annex B sz("LLU") < Leb128Bytes("LLU") + frame_unit_size("LLU")\n", sz, Leb128Bytes, frame_unit_size));
+			return GF_NON_COMPLIANT_BITSTREAM;
+		}
+		GF_LOG(GF_LOG_DEBUG, GF_LOG_CONTAINER, ("[AV1] Annex B    frame unit detected (size "LLU")\n", frame_unit_size));
+		sz -= Leb128Bytes + frame_unit_size;
+
+		while (frame_unit_size > 0) {
+			ObuType obu_type;
+			u64 obu_length = leb128(bs, &Leb128Bytes);
+			if (frame_unit_size < Leb128Bytes + obu_length) {
+				GF_LOG(GF_LOG_ERROR, GF_LOG_CODING, ("[AV1] Annex B frame_unit_size("LLU") < Leb128Bytes("LLU") + obu_length("LLU")\n", frame_unit_size, Leb128Bytes, obu_length));
+				return GF_NON_COMPLIANT_BITSTREAM;
+			}
+			GF_LOG(GF_LOG_DEBUG, GF_LOG_CONTAINER, ("[AV1] Annex B        OBU detected (size "LLU")\n", obu_length));
+			u64 pos = gf_bs_get_position(bs);
+			frame_unit_size -= Leb128Bytes;
+
+			e = gf_media_aom_av1_parse_obu(bs, &obu_type, &obu_length, state);
```

Comment:
If a new `obu_inner_length` is passed to `gf_media_aom_av1_parse_obu`, a check on whether the values inside the OBU header (if any) and outside (in the Annex B) are the same could be added.